packages/itwinui-react/src/core/Table/Table.tsx

```diff
@@ -578,18 +579,42 @@ export const Table = <
     );
   }, [data, getSubRows]);
 
-  const [rowHasParent] = React.useState(() => new WeakSet<T>());
-
   const getSubRowsWithSubComponents = React.useCallback(
-    (originalRow: T) => {
-      if (!rowHasParent.has(originalRow)) {
-        rowHasParent.add(originalRow);
-        return [originalRow];
+    (originalRow: T, relativeIndex: number) => {
+      const symbol = Symbol.for('iui-id');
+
+      // Check if the current row is a sub-row that holds data for the according sub-component.
+      if (originalRow[symbol as keyof typeof originalRow]) {
+        return [];
       }
 
-      return (originalRow.subRows as T[]) || [];
+      // If rows have sub-components (generated as sub-rows), we need to return that sub-rows. Otherwise, row selection will get into a loop as it needs the row ID to set the sub-component to be selected.
+      // If the main row has not had any sub-rows, we need an object that holds a Symbol as a form of flag to prevent infinite loop.
+      return (
+        (originalRow.subRows as T[]) ?? [
```

Comment:
The reason why I check if `subRows` are present is because when the row is selected, we already have all sub-rows generated. The react-table `useRowSelect` hook keeps getting through all sub-rows via IDs using this `getSubRowsWithSubComponents` function to set main row and all nested sub-rows selected (https://github.com/TanStack/table/blob/v7/src/plugin-hooks/useRowSelect.js#L166C8-L167C71). However, since we 
 return the new object with the spread `originalRow`, it kept accessing the same `row.id` over again. Therefore, I had to return the sub-rows (our sub-component data) so it could set the sub-component row selected and get out of the loop.